File: client/Assets/Scripts/Network/GameClient.cs
// NexusIsekai Client - GameClient.cs
// Quản lý kết nối TCP đến game server, xử lý protocol length-prefix 4 byte

using System;
using System.Net.Sockets;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Concurrent;
using UnityEngine;

namespace NexusIsekai.Network
{
    public class GameClient : MonoBehaviour
    {
        // Singleton
        public static GameClient Instance { get; private set; }

        [Header("Server Config")]
        public string serverHost = "127.0.0.1";
        public int    serverPort = 7777;

        [Header("Network Settings")]
        public float reconnectDelay   = 3f;
        public int   receiveBufferSize = 65536;

        // Trạng thái kết nối
        public enum ConnectionState { Disconnected, Connecting, Connected }
        public ConnectionState State { get; private set; } = ConnectionState.Disconnected;

        public bool IsConnected => State == ConnectionState.Connected;

        // Events — dispatch trên main thread qua packet queue
        public event Action                  OnConnected;
        public event Action<string>          OnDisconnected;   // reason
        public event Action<short, byte[]>   OnPacketReceived; // opcode, payload

        private TcpClient   _tcp;
        private NetworkStream _stream;
        private Thread      _receiveThread;
        private CancellationTokenSource _cts;

        // Queue để dispatch packet về main thread (Unity không thread-safe)
        private readonly ConcurrentQueue<(short opcode, byte[] payload)> _packetQueue
            = new ConcurrentQueue<(short, byte[])>();

        // ─────────────────────────────────────────
        // Unity lifecycle
        // ─────────────────────────────────────────

        private void Awake()
        {
            if (Instance != null && Instance != this) { Destroy(gameObject); return; }
            Instance = this;
            DontDestroyOnLoad(gameObject);
        }

        private void Update()
        {
            // Dispatch packet queue về main thread
            while (_packetQueue.TryDequeue(out var pkt))
            {
                try { OnPacketReceived?.Invoke(pkt.opcode, pkt.payload); }
                catch (Exception e) { Debug.LogError($"Packet handler error: {e}"); }
            }
        }

        private void OnApplicationQuit() => Disconnect("App quit");

        // ─────────────────────────────────────────
        // Connect / Disconnect
        // ─────────────────────────────────────────

        public async Task ConnectAsync()
        {
            if (State != ConnectionState.Disconnected) return;

            State = ConnectionState.Connecting;
            _cts  = new CancellationTokenSource();

            try
            {
                _tcp = new TcpClient { ReceiveBufferSize = receiveBufferSize, NoDelay = true };
                await _tcp.ConnectAsync(serverHost, serverPort);
                _stream = _tcp.GetStream();

                State = ConnectionState.Connected;
                Debug.Log($"[GameClient] Connected to {serverHost}:{serverPort}");

                // Bắt đầu receive thread
                _receiveThread = new Thread(ReceiveLoop) { IsBackground = true };
                _receiveThread.Start();

                OnConnected?.Invoke();
            }
            catch (Exception e)
            {
                State = ConnectionState.Disconnected;
                Debug.LogError($"[GameClient] Connect failed: {e.Message}");
                OnDisconnected?.Invoke(e.Message);
            }
        }

        public void Disconnect(string reason = "")
        {
            if (State == ConnectionState.Disconnected) return;

            _cts?.Cancel();
            State = ConnectionState.Disconnected;

            try { _stream?.Close(); _tcp?.Close(); } catch { }
            _stream = null;
            _tcp    = null;

            Debug.Log($"[GameClient] Disconnected: {reason}");
            // Dispatch disconnect về main thread qua Update
            _packetQueue.Enqueue((0, null)); // sentinel
        }

        // ─────────────────────────────────────────
        // Send
        // ─────────────────────────────────────────

        /// <summary>
        /// Gửi packet: [4 byte big-endian length][2 byte opcode][payload]
        /// Length = số byte SAU length field = opcode + payload
        /// </summary>
        public void Send(PacketBuilder pkt)
        {
            if (!IsConnected) return;

            try
            {
                byte[] data = pkt.Build();
                lock (_stream) { _stream.Write(data, 0, data.Length); }
            }
            catch (Exception e)
            {
                Debug.LogError($"[GameClient] Send error: {e.Message}");
                Disconnect("Send error");
            }
        }

        // ─────────────────────────────────────────
        // Receive loop (chạy trên background thread)
        // ─────────────────────────────────────────

        private void ReceiveLoop()
        {
            byte[] lenBuf = new byte[4];

            try
            {
                while (!_cts.IsCancellationRequested)
                {
                    // Đọc 4 byte length (big-endian)
                    if (!ReadFully(lenBuf, 4)) break;

                    int bodyLen = (lenBuf[0] << 24) | (lenBuf[1] << 16)
                                | (lenBuf[2] <<  8) |  lenBuf[3];

                    if (bodyLen < 2 || bodyLen > 1_048_576) // max 1MB
                    {
                        Debug.LogError($"[GameClient] Invalid packet length: {bodyLen}");
                        break;
                    }

                    // Đọc body
                    byte[] body = new byte[bodyLen];
                    if (!ReadFully(body, bodyLen)) break;

                    // Parse opcode (2 byte big-endian, đầu body)
                    short opcode = (short)((body[0] << 8) | body[1]);
                    byte[] payload = new byte[bodyLen - 2];
                    Buffer.BlockCopy(body, 2, payload, 0, payload.Length);

                    _packetQueue.Enqueue((opcode, payload));
                }
            }
            catch (Exception e)
            {
                if (!_cts.IsCancellationRequested)
                    Debug.LogError($"[GameClient] Receive error: {e.Message}");
            }
            finally
            {
                if (State == ConnectionState.Connected)
                {
                    State = ConnectionState.Disconnected;
                    // Enqueue sentinel để main thread biết disconnect
                    _packetQueue.Enqueue((PacketOpcode.S2C_KICK, null));
                }
            }
        }

        private bool ReadFully(byte[] buf, int count)
        {
            int offset = 0;
            while (offset < count)
            {
                if (_cts.IsCancellationRequested) return false;
                int n = _stream.Read(buf, offset, count - offset);
                if (n == 0) return false; // server đóng kết nối
                offset += n;
            }
            return true;
        }
    }
}

File: client-android/app/src/main/java/com/nexusisekai/net/GameClient.java
package com.nexusisekai.net;

import android.util.Log;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * GameClient — TCP connection tới game server.
 * Protocol: [4-byte big-endian length][2-byte opcode][payload...]
 *
 * Singleton, gọi GameClient.getInstance() từ bất kỳ đâu.
 */
public class GameClient {

    private static final String TAG = "GameClient";
    private static GameClient INSTANCE;

    public static synchronized GameClient getInstance() {
        if (INSTANCE == null) INSTANCE = new GameClient();
        return INSTANCE;
    }

    private Socket           socket;
    private DataInputStream  dis;
    private DataOutputStream dos;
    private volatile boolean running   = false;
    private volatile boolean connected = false;

    // Send queue — gửi từ bất kỳ thread nào
    private final BlockingQueue<byte[]> sendQueue = new LinkedBlockingQueue<>(1000);

    // Executor cho read/write threads
    private final ExecutorService executor = Executors.newFixedThreadPool(2);

    // Listener nhận packet dispatched
    private PacketListener listener;

    public interface PacketListener {
        void onPacket(short opcode, byte[] payload);
        void onDisconnected(String reason);
    }

    public void setListener(PacketListener l) { this.listener = l; }

    // ─── Connect ─────────────────────────────────────────────

    public void connect(String host, int port) {
        executor.submit(() -> {
            try {
                socket = new Socket();
                socket.connect(new InetSocketAddress(host, port), 10_000);
                socket.setTcpNoDelay(true);
                socket.setKeepAlive(true);
                dis = new DataInputStream(socket.getInputStream());
                dos = new DataOutputStream(socket.getOutputStream());
                running   = true;
                connected = true;
                Log.i(TAG, "Connected to " + host + ":" + port);
                executor.submit(this::readLoop);
                executor.submit(this::writeLoop);
            } catch (IOException e) {
                Log.e(TAG, "Connect failed: " + e.getMessage());
                if (listener != null) listener.onDisconnected(e.getMessage());
            }
        });
    }

    public void disconnect() {
        running   = false;
        connected = false;
        sendQueue.clear();
        try { if (socket != null) socket.close(); } catch (IOException ignored) {}
    }

    public boolean isConnected() { return connected; }

    // ─── Read loop ────────────────────────────────────────────

    private void readLoop() {
        try {
            while (running) {
                int bodyLen = dis.readInt();
                if (bodyLen < 2 || bodyLen > 1_048_576) throw new IOException("Bad packet length: " + bodyLen);
                short opcode    = dis.readShort();
                int payloadLen  = bodyLen - 2;
                byte[] payload  = new byte[payloadLen];
                if (payloadLen > 0) dis.readFully(payload);
                if (listener != null) listener.onPacket(opcode, payload);
            }
        } catch (IOException e) {
            if (running) {
                running = connected = false;
                Log.w(TAG, "Read error: " + e.getMessage());
                if (listener != null) listener.onDisconnected(e.getMessage());
            }
        }
    }

    // ─── Write loop ───────────────────────────────────────────

    private void writeLoop() {
        try {
            while (running) {
                byte[] pkt = sendQueue.poll(500, java.util.concurrent.TimeUnit.MILLISECONDS);
                if (pkt != null) {
                    dos.write(pkt);
                    dos.flush();
                }
            }
        } catch (Exception e) {
            if (running) {
                running = connected = false;
                if (listener != null) listener.onDisconnected(e.getMessage());
            }
        }
    }

    // ─── Send ─────────────────────────────────────────────────

    public void send(PacketWriter pw) {
        if (!connected) { Log.w(TAG, "Not connected, dropping packet"); return; }
        sendQueue.offer(pw.build());
    }

    public void sendOpcode(short opcode) { send(new PacketWriter(opcode)); }
}
